Microsoft.AVS.VVOLS/Microsoft.AVS.VVOLS.psm1

```diff
@@ -0,0 +1,370 @@
+        [ValidateNotNull()]
+        [String]
+        $ProviderName
+    )
+
+    $VasaProvider = Get-VasaProvider -Name $ProviderName -ErrorAction Ignore
+    if (-not $VasaProvider) {
+        throw "Unable to remove a VASA provider. Vasa provider '$ProviderName' does not exist."
+    }
+
+    Remove-VasaProvider -Provider $VasaProvider -Confirm:$false -ErrorAction Stop | Out-Null
+
+    Write-Host "VASA provider $ProviderName removed successfully."
+}
+
+
+<#
+    .SYNOPSIS
+     Creates a new vVol Storage Policy
+
+    .PARAMETER PolicyConfigJsonString
+     Storage policy in Json string. The command will traverse through the SPBM rules and creat a storage policy containing all of the rules.
+     The example string is as below:
+    {
+        "SchemaVersion": "1.0.0",
+        "Vendor": "Pure Storage",
+        "PolicyName": "",
+        "PolicyDescription": "",
+        "SpbmRules": {
+            "com.purestorage.storage.policy.FlashArrayGroup":["fa1","fa2"],
+            "com.purestorage.storage.replication.LocalSnapshotPolicyCapable": true,
+            "com.purestorage.storage.replication.LocalSnapshotInterval":"00:00:00"
+            ...
+        }
+    }
+
+    .EXAMPLE
+     New-VvolStoragePolicy -PolicyConfigJsonString $policyConfigJsonString
+
+    .INPUTS
+     vCenter storage policy in Json string
+
+    .OUTPUTS
+     None.
+#>
+function New-VvolStoragePolicy {
+    [CmdletBinding()]
+    [AVSAttribute(10, UpdatesSDDC = $false)]
+    Param(
+        [Parameter(
+            Mandatory = $true,
+            HelpMessage = 'Storage policy in Json string')]
+        [ValidateNotNull()]
+        [String]$PolicyConfigJsonString
+    )
+    # Convert JSON to PowerShell object
+    $PolicyConfig = ConvertFrom-Json $PolicyConfigJsonString
+    # Extract values from the object
+    $Vendor = $PolicyConfig.Vendor
+    $PolicyName = $PolicyConfig.PolicyName
+    $PolicyDescription = $PolicyConfig.PolicyDescription
+    $SchemaVersion = $PolicyConfig.SchemaVersion
+
+    if ($SchemaVersion -gt "1.0.0") {
+        throw "Unable to create a storage policy. Schema version $SchemaVersion is not supported."
+    }
+
+    # Create SPBM rules dictionary
+    $rules = @()
+    $PolicyConfig.SpbmRules.PSObject.Properties | ForEach-Object {
+        $RuleName = $_.Name
+        $RuleValue = $_.Value
+        if ($Vendor -eq "Pure Storage") {
+            $ValueType = (Get-SpbmCapability -Name $RuleName).ValueType.Name
+            if ($RuleValue -is [int64] -and $ValueType -eq "int32") {
+                $RuleValue = [int]$RuleValue
+            }
+            if ($ValueType -eq "TimeSpan") {
+                $RuleValue = [TimeSpan]::Parse($RuleValue)
+            }
+        }
+        $rules += New-SpbmRule -Capability (Get-SpbmCapability -Name $RuleName) -Value $RuleValue
+    }
+
+    Write-Host "Creating policy $PolicyName for vendor $Vendor..."
+    $Ruleset = New-SpbmRuleSet -AllOfRules $rules
```

Comment:
How do we make sure that the policy is only applicable to VVOL objects?

Reply:
We cannot do this a t the time of creating the policy.